device/pci_device.cpp

```diff
@@ -0,0 +1,431 @@
+    static const char sys_pattern[] = "/sys/bus/pci/devices/%04x:%02x:%02x.%u/%s";
+    char buf[sizeof(sys_pattern) + 10];
+
+    // revision pattern = "/sys/bus/pci/devices/%04x:%02x:%02x.%u/revision"
+    std::snprintf(buf, sizeof(buf), sys_pattern, pcie_domain, pcie_bus, pcie_device, pcie_function, "revision");
+
+    std::ifstream revision_file(buf);
+    std::string revision_string;
+    if (std::getline(revision_file, revision_string)) {
+        this->pcie_device_id = device_info.out.device_id;
+        this->pcie_revision_id = std::stoi(revision_string, nullptr, 0);
+    } else {
+        TT_THROW("Revision ID /sys/ read failed for device: ", this->device_id);
+    }
+
+    // Get NUMA node from sysfs
+    // numa node pattern = "/sys/bus/pci/devices/%04x:%02x:%02x.%u/numa_node"
+    std::snprintf(buf, sizeof(buf), sys_pattern, pcie_domain, pcie_bus, pcie_device, pcie_function, "numa_node");
+
+    std::ifstream num_node_file(buf);
+    std::string numa_node_string;
+    if (std::getline(num_node_file, numa_node_string)) {
+        this->numa_node = std::stoi(numa_node_string, nullptr, 0);
+    } else {
+        TT_THROW("Numa node /sys/ read failed for device: ", this->device_id);
+    }
+}
+
+// Open a unique device_id per host memory channel (workaround for ttkmd < 1.21 support for more than 1 pin per fd)
+void TTDevice::open_hugepage_per_host_mem_ch(uint32_t num_host_mem_channels) {
+    for (int ch = 0; ch < num_host_mem_channels; ch++) {
+        log_debug(LogSiliconDriver, "Opening device_fd_per_host_ch device index: {} ch: {} (num_host_mem_channels: {})", device_id, ch, num_host_mem_channels);
+        int device_fd_for_host_mem = find_device(device_id);
+        if (device_fd_for_host_mem == -1) {
+            throw std::runtime_error(std::string("Failed opening a host memory device handle for device ") + std::to_string(device_id));
+        }
+        device_fd_per_host_ch.push_back(device_fd_for_host_mem);
+    }
+}
+
+tt::ARCH TTDevice::get_arch() const {
+    return arch;
+}
+
+template<typename T>
+T* TTDevice::get_register_address(std::uint32_t register_offset) {
+    void *reg_mapping;
+    if (system_reg_mapping != nullptr && register_offset >= system_reg_start_offset) {
+        register_offset -= system_reg_offset_adjust;
+        reg_mapping = system_reg_mapping;
+    } else if (bar0_wc != bar0_uc && register_offset < bar0_wc_size) {
+        reg_mapping = bar0_wc;
+    } else {
+        register_offset -= bar0_uc_offset;
+        reg_mapping = bar0_uc;
+    }
+    return reinterpret_cast<T*>(static_cast<uint8_t*>(reg_mapping) + register_offset);
+}
+
+void TTDevice::write_block(uint64_t byte_addr, uint64_t num_bytes, const uint8_t* buffer_addr) {
+    void *dest = nullptr;
+    if (bar4_wc != nullptr && byte_addr >= BAR0_BH_SIZE) {
+        byte_addr -= BAR0_BH_SIZE;
+        dest = reinterpret_cast<uint8_t *>(bar4_wc) + byte_addr;
+    }else {
+        dest = get_register_address<uint8_t>(byte_addr);
```

Comment:
I think the 'register' part of `get_register_address` is a bit misleading here.. if I understand the code correctly, it can end up writing to a register exposed in a PCIe BAR, or write through an inbound TLB window (to what might be either memory or a register).

It is probably worth making a distinction between these scenarios at the interface level.